benchmark/benchmark-atomic.cpp

```diff
@@ -0,0 +1,254 @@
+struct ExecPolicyOMP {
+    using policy = RAJA::omp_for_exec;
+    using sync = RAJA::policy::omp::omp_synchronize;
+    static std::string PolicyName() {
+        std::stringstream ss;
+        ss << "OpenMP execution";
+        return ss.str();
+    }
+};
+
+/// Functor wrapping the desul implementation.  Wrapping the desul call ensure an identical signature with
+/// RAJA's implementations.  Wrapping the call in an functor allows simple type deduction for printing
+/// from within the benchmark.
+template<typename T, typename DesulAtomicSignature<T>::type atomic_impl>
+struct atomicWrapperDesul {
+    /// Call operator overload template that allows invoking DESUL atomic with a (int*)(T*, T) signature
+    RAJA_DEVICE T operator()(T * acc, T value) const {
+        return atomic_impl(acc, value, raja_default_desul_order{},
+                        raja_default_desul_scope{});
+    }
+};
+
+template<typename T>
+class IsDesul : public std::false_type {};
+
+template<typename T, typename DesulAtomicSignature<T>::type atomic_impl>
+class IsDesul<atomicWrapperDesul<T, atomic_impl>> : public std::true_type {};
+
+template<typename AtomicImplType>
+std::string GetImplName (const AtomicImplType& impl) {
+    if (IsDesul<AtomicImplType>::value) {
+        return "Desul atomic";
+    } else {
+        return "RAJA atomic";
+    }
+}
+
+/// Helper to deduce if atomic implementation is using desul under the hood
+//bool isDesul(typename RajaAtomicSignature<AtomicType>::signature atomic_impl) {
+//    bool is_desul =
+//        atomic_impl == atomicWrapperDesul<AtomicType, typename GPUAtomic::policy, desul::atomic_fetch_add> ||
+//        atomic_impl == atomicWrapperDesul<AtomicType, typename GPUAtomic::policy, desul::atomic_fetch_sub> ||
+//        atomic_impl == atomicWrapperDesul<AtomicType, typename GPUAtomic::policy, desul::atomic_fetch_>
+//
+//}
+
+template <class ExecPolicy, typename AtomicType, bool test_array, typename AtomicImplType>
+void TimeAtomicOp( AtomicImplType atomic_impl, uint64_t num_iterations = 2, int array_size = 100) {
+    RAJA::Timer timer;
+
+    // Allocate memory
+    AtomicType* device_value = nullptr;
+    int len_array = test_array ? array_size : 1;
+    if (IsGPU<typename ExecPolicy::policy>::value) {
+        AllocateAtomicDevice(&device_value, len_array);
+    } else {
+        device_value = new AtomicType [len_array];
+    }
+
+    for (int i = 0; i < num_iterations; ++i) {
+        // Don't time the first kernel launch
+        if (i == 1) {
+            timer.start();
+        }
+        RAJA::forall<typename ExecPolicy::policy>(RAJA::RangeSegment(0, N),
+        [=] RAJA_HOST_DEVICE(int tid)  {
+            if (test_array) {
+                atomic_impl(&(device_value[tid % array_size]), AtomicType(1));
+            } else {
+                atomic_impl(device_value, AtomicType(1));
+            }
+        });
+    }
+    timer.stop();
+    RAJA::synchronize<typename ExecPolicy::sync>();
```

Comment:
You will want to synchronize before stopping the timer though to ensure the work is finished.